[RFC] Device: generalizing and improving control features (sync, lock and timeout)

tbursztyka opened the pull request:
Part of #22941

(This RFC is now presenting ideas, and is currently not the final implementation which will evolve through comments until it reaches a consensus)

Currently synchronizing calls on interrupt based devices is a feature implemented by each and every drivers. This becomes worse about concurrent access mitigation, where many drivers do not implement anything.

This RFC tries then to provide an approach on generalizing both synchronization and concurrent access mitigation to all devices. It also adds a new feature: setting a timeout on calls. (currently working only on interrupt based devices, adding it for polling based ones is entirely possible but to simplify this RFC was put aside)

All - but the sync part - is configurable through Kconfig:
- concurrent access mitigation can be disabled (on targets where devices are not accessed by +2 threads for instance)
- timeouts are by default disabled

**Important notice:** Many driver are by default always working synchronously and thus would not make any usage of the sync part of this RFC. I did not take this into account for now. It would be entirely possible to save memory by having a parameter on the device instance macros (thus why I did not try to play with this yet, due to the amount of macro call to be fixed) to tell about that requirement, and thus save a sensible amount of memory since struct k_sem is not tiny.
We could do the same about the lock on concurrent access, if relevant. Having a NULL pointer instead of struct k_sem pointer would be trivial to manage.

[Note] This RFC if built on top of #23407 

### Idea and use cases

**Synchronization**

All device driver synchronous API call that are implemented on interrupt basis in the driver would use is:
- api_call()
-> configure and start the interrupt based work
-> k_sem_take(sync_semaphore) <--- it would wait until the interrupt based work is finished and runs a k_sem_give(sync_semaphore)

A large amount of device drivers already use that scheme. But they are currently implementing such synchronization on their own. Idea is to generalize it.  (See the important notice above, however)

**Concurrent access**

All device driver API calls, no exception (whether these calls are synchronous or not).
This is a basic yet effective way to protect the device driver resource.
(See the important notice above also)

**Timeout**

Currently, only a few device driver API provide a timeout and only on certain call. But this will be insufficient if we seek for robustness. A tiny hw issue could just lock the whole device and there would be no way for recovering it.

Idea is to simply add such generic timeout on every call. It could be using a default setting as well as a dedicated one. 
Note however that the current RFC is simply setting such timeout on the sync semaphore. It would be easy to run it separately though.

Commit:
device: Introduce a generic context synchronization mechanism

This adds a synchronization semaphore to generalize call synchronization
among all device drivers. This will permit to avoid for drivers
maintainers to implement their own solution.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Add an optional timeout on generic locking mechanism

This feature could be used to avoid being locked too long on a call, but
would require the caller to manage such error.

On a broader point a view, this is the first and necessary iteration to
get to a generic call timeout which will avoid to be locked if the
hardware fails (a SPI transaction that never finishes for instance).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/kernel: Add a test for the device lock timeout feature

2 concurrent calls will lead to the second one getting -EAGAIN as an
error returned.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/kernel: Move current device test to a new directory

There will be more specific tests inside tests/kernel/device

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/kernel: Add a test for the device control part

This is testing generic locking and syncing on device calls.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Add a timeout feature on API call

By default this is disabled, to follow legacy behavior.

If enabled, this will let the possibilyt to set a timeout on device API
call (either general via CONFIG_DEVICE_CALL_TIMEOUT_VALUE, or specific
ones).

Currently the feature is meant to work only on interrupt based API
calls. To get this feature working on polling-based calls, it would
require having a dedicated k_delayed_work in kernel/device.c, configured
to trigger an actual device call cancelation on the soonest on-going
call timeout.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on kernel/Kconfig.device:
```diff
@@ -0,0 +1,24 @@
+	help
+	  This will set the generic timeout for device locking.
+	  By default -1 is set (K_FOREVER), any positive value is valid.
+	  Make sure to select a suitable value for all the devices.
```
Could this value be device specific? Perhaps we could have it in DEVICE_DEFINE(), then we could avoid global value that might not be suitable for all devices.

tbursztyka replied:
Sure, like the presence of a sync or lock actually. I did not push any change of the macros yet (or then there would be a big patch changing the macros everywhere it's being called), so we can settle on what would be exposed - or not - as parameters there.

nordic-krch commented on kernel/device.c:
```diff
@@ -177,3 +179,45 @@ void device_busy_clear(struct device *busy_dev)
+		(dev - __device_start);
+	u32_t status;
+
+	k_sem_take(&dc->sync, K_FOREVER);
```
forever here is risky. What if call is never completed.

tbursztyka replied:
This is improved afterwards with a timeout. However, note that current sync semaphores in drivers afaik always use K_FOREVER. (which is bad indeed)

nordic-krch commented on kernel/device.c:
```diff
@@ -177,3 +179,45 @@ void device_busy_clear(struct device *busy_dev)
+int device_lock_timeout(struct device *dev, k_timeout_t timeout)
+{
+#ifdef CONFIG_DEVICE_CONCURRENT_ACCESS
+	struct device_context *dc =
```
i am not convinced that we want to have access control context for all devices. It takes ram and for many devices it is not needed because there is a single user or other methods to maintain control. I'm also pushing for allowing using some devices from interrupt context (i2c/spi) where access control is maintained using queue operation manager (#24781). Imo, it would be better to have pointer to that context in the device structure. null meaning no access control.

tbursztyka replied:
Yes, I noted that in this RFC message, see "Important notice". 

If there is no resistance in generalizing this concepts, I will improve the RFC so it will be possible to allocate sync/lock semaphores per-instance basis and not to all.

pabigot commented on kernel/device.c:
```diff
@@ -23,6 +23,8 @@ extern const struct init_entry __init_SMP_start[];
 extern struct device __device_start[];
 extern struct device __device_end[];
 
+extern struct device_context __device_context_start[];
```
This approach seems to disallow a specific driver implementation from extending the base context with its own additional context.  Having `struct device_context *context` in `struct_device` would be more general, allowing `CONTAINER_OF` to upcast to access additional information.

This could complicate the initialization macros; OTOH the context could be put into BSS and the sync/lock fields initialized on boot by the init loop.

tbursztyka replied:
struct device's data attribute is the device driver dedicated context.
I thought we could merge both, as I did for dedicated api exposure in net stack for 15.4 (othe l2 have been using the same trick then). All specialized api have to get a net_if_api attribute in first place.

But it could be done afterwards (or before, either way), in order not to do too much changes in this PR. It's somehow 2 different things. Thus also why I did not add yet another pointer in struct device.

pabigot reviewed (commented):
While I'm not against the idea of sharing a common lock and sync mechanism, I'm not ready to commit to it being a good idea.  Does it really generalize to all devices?

I think I'd rather see this approach prototyped in one or two subsystems before being promoted to all devices.

tbursztyka commented:
>  Does it really generalize to all devices?

I could try to compile a list. Those who are of polling-based implementation will not use the sync semaphore for instance, all other ones will. Thus why the device macro should let the driver maintainer control if such sync semaphore should be allocated or not. The lock is a different story, so far I cannot see any driver that would not need it on a multi-thread build. Nothing prevents a thread to collide with another when accessing a device instance, the API calls are not atomic. Would be nice to track driver's users at built time, so it would open for some optimizations, but I don't see this really possible (not with current device_get_binding() at the very least)

> I think I'd rather see this approach prototyped in one or two subsystems before being promoted to all devices.

drivers/spi/spi_context.h used by all spi drivers is basically doing that. 
I could try to replace it with this current proposal.

pabigot commented:
In general API calls may not be atomic, but certainly for some drivers it's important that they be callable from ISRs (#6184, others from #18970).  GPIO in particular, probably COUNTER (#17014).  (At some point we need to go back and update all the API so that we can what's safe to call from interrupt context, not only by documentation but at runtime for when it differs between devices.)

I'm also skeptical of supporting timeouts for every operation (though if we do we might get `isr-ok` by using `K_NO_WAIT`).  That's a change with global impact, as it implies that it must always be possible for the application/driver to do something to recover if the timeout fires.  In some applications the appropriate approach may very well be the existing behavior: deadlock, allowing the watchdog to fire and hard-reset the system.

tbursztyka commented:
> That's a change with global impact, as it implies that it must always be possible for the application/driver to do something to recover if the timeout fires. In some applications the appropriate approach may very well be the existing behavior: deadlock, allowing the watchdog to fire and hard-reset the system.

Wait, such timeout will not be enabled by default, and it never will. But at least zephyr will be able to propose both behaviors.

I don't see much trouble with this PR and #6184, it would be quite easy to no lock/release a device on *_isr type of API. That said, I fully agree: we need to solve #6184.

pabigot commented:
Reviewing this PR will be difficult because it's adding three features (concurrency control, internal synchronization, and timeouts), all of which could be discussed to death.

I'd like to focus some discussion on the method used to prevent current access, i.e. the `device_lock{,_timeout}()` API.  The proposal here is to use a semaphore.

Is this the right choice?

Somewhere in this project there's a comment to the effect of:

> If you don't need recursion or priority inheritance, use a semaphore [instead of a mutex].

Why would we not want priority inheritance for generic driver access?

I think there was also some concern that a mutex requires more space than a semaphore.  I'm finding that a mutex on ARM32 is 20 bytes.  A semaphore is 24 bytes if `CONFIG_POLL=y`, otherwise it's 16 bytes.

Since signals are the only mechanism for implementing async operations from user mode, most async APIs that aren't restricted to privileged code will require `CONFIG_POLL=y` to receive notification when the operation completes, so I believe 24 bytes is a better estimate of the memory cost of a semaphore.

As mutexes take less memory and provide an important feature, such I'd like to understand why a semaphore is a better choice for generic driver concurrency control than a mutex.

tbursztyka commented:
> Reviewing this PR will be difficult because it's adding three features (concurrency control, internal synchronization, and timeouts), all of which could be discussed to death.

As it is an RFC, it all about the idea it brings, not about proper atomic patches that we found in normal PR that are meant to be merged. Provided patches are quite short, I believe contributors can grasp the idea by looking at these. Then, **and only then**, if these ideas are judged worth to be properly implemented then it will be done via a proper PR. So far, it has not brought much interest, let's still wait.

> Why would we not want priority inheritance for generic driver access?

On interrupt based drivers (which is the majority of drivers), that would be pretty much useless since the whole driver processing task would be done through the isr and thus via the interrupt's priority, not the thread.
The question is more like: do we need thread priority inheritance anywhere there?

> As mutexes take less memory and provide an important feature, I'd like to understand why a semaphore is a better choice for generic driver concurrency control than a mutex.

How would you unlock a mutex in ISR context?


That said, about memory occupation, I always found it weird to have semaphore poll event type. It's been used in one place in the whole codebase (where it could be easily replaced by something else). Perhaps it is worth to take a look at that. 32 bits for count & limit is rather extremely large as well. Maybe some room for optimization there if required.

pabigot commented:
> How would you unlock a mutex in ISR context?

Per #25676, clearly, you wouldn't.

However I don't recall a case where I could release the in-use lock for a driver from an ISR.  When whatever caused a suspend completes there's often some cleanup that the driver has to perform before it's safe to allow something else to use the driver.  And I would want that clean-up code to take priority over other things if there's a higher-priority task waiting to use the device.

tbursztyka commented:
> However I don't recall a case where I could release the in-use lock for a driver from an ISR.

I was on both sync and lock, nevermind. The lock is always taken/released by the caller's thread. 

> I would want that clean-up code to take priority over other things if there's a higher-priority task waiting to use the device.

That's a confusing idea. I don't see how a mutex (or a semaphore actually) would do anything about it here.

Both work the same when dealing with their wait queue, a waiting thread is by default not ready to be scheduled until the unlock happens anyway, which one cannot be made before the clean-up. Which clean-up happens: in polling based driver it's in caller's thread context at all time, in interrupt based mode it's usually directly in isr context when final interrupt has been caught.

Clean-up is generally short (if not really short), I don't see the benefit to generalize a re-prioritization of the caller's thread to run asap in order for the next high prio thread in the wait queue (if any) to get the device afterwards (how likely this is going to happen vs how much processing time this strategy would require). Actually, it seems like a bad idea to re-prioritize a thread owning a mutex (see k_thread_priority_set() documentation).

Commit:
device: Introduce device context for concurrent access mitigation

This introduce a generic device concurrent access protection.

Such feature already exist, scattered either through some driver
domain API - SPI for instance - or most of the time per-driver when the
maintainer has a good understanding on what to do. But not everywhere.
Many drivers are not protected against concurrent access. Generalizing
this will remove the burden for driver maintainers to do that on their
own, but also will add it to all the drivers which are lacking this.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Make the lock/release function ISR aware

If it is called from an ISR, it will directly return successfully.

Note: This is separated from original lock feature addition for a
reason. Only and handful of device APIs are ISR proof (gpio for
instance), all others are not. Should the lock be generalized to all
device APIs, this would at least prevent ISR proof APIs to misbehave
compared to what they already do.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Make the synchronization ISR aware.

Again to avoid misbehavior in ISR context for ISR ready device API
calls.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Got some time to split patches so it gets one feature added per-patch. More commits, but if that can help people to get the idea.

tbursztyka commented:
> >   I would want that clean-up code to take priority over other things if there's a higher-priority task waiting to use the device.
> That's a confusing idea. I don't see how a mutex (or a semaphore actually) would do anything about it here.

Ok I see now. I missed the kernel/mutex.c line 160.
Which explains why k_thread_priority_set() should not be used on a thread owning a mutex (documentation could be much clearer there then).

tbursztyka commented:
@pabigot There is one issue with using mutexes as a lock object: SPI and ADC provide async calls (the result notification of those calls). With a semaphore, a caller cannot subsequently raise 2+  of these async call. With a mutex, since there is no limit it becomes a problem.

pabigot commented:
> With a semaphore, a caller cannot subsequently raise 2+ of these async call. With a mutex, since there is no limit it becomes a problem.

That's a good point.  So the idiom if we use mutex should include a flag that's checked at all public entry-points and rejects with `-EBUSY` if the mutex was taken while there's already a task in progress.  There may be an alternative solution that involves a non-recursive mutex, but I think that would introduce new and worse problems.

#24781 is intended to support uses where there are multiple pending operations from the same or different threads, addressing a gap in the current async solution (any operation requested while the device is busy will block, which is not expected for async functions).  There it's the queued operation manager that would be preventing multiple operations from being invoked simultaneously.  (Of course you can't use the direct async API when a queued operation manager is controlling the device because it violates the expectation that one thing controls access.)

pabigot commented:
Related to implementing internal synchronization in the core driver rather than each driver: Several drivers (particularly flash drivers) need to work with `CONFIG_MULTITHREADING=n` so that use within a bootloader doesn't pull in a bunch of code that isn't needed.

See #26372 for example.

Similarly at least GPIO and UART, and some COUNTER, must have API that works from interrupt mode, so not every API function can require external synchronization through a mutex or semaphore.

tbursztyka commented:
> See #26372 for example.

Good point, did not know about it.

> Similarly at least GPIO and UART, and some COUNTER, must have API that works from interrupt mode, so not every API function can require external synchronization through a mutex or semaphore.

Yes we need to settle on #6184
About this and the  proposed API, I'll extend DEVICE_DEFINE() in order to tell whether a sync and/or lock are necessary. So drivers that do not require it will not get any (instead of blindly allocating for all)

joerchan commented:
Not sure why Github decided to remove so many people, I only wanted to remove myself.
